### ISSUE TITLE
Inconsistent weight calculation approach for disputes in `paras_inherent`

### ISSUE THREAD

**Issue by tdimitrov** [paritytech/polkadot-sdk#2122]
`process_inherent_data` receives bitfields, backed candidates and disputes as input and produces `ParachainsInherentData` from them. The function is not trivial but generally it runs some sanitization code, applies weight limit and processes each input. 

To avoid creating overweight blocks [`apply_weight_limit`](https://github.com/paritytech/polkadot-sdk/blob/f50054cffba4df6b25a16dafdbc66d582e419355/polkadot/runtime/parachains/src/paras_inherent/mod.rs#L445) limits the bitfields and backed candidates that are included in a block. By limit I mean drop candidates/bits until they fit in the limit. How this is done is not important for this issue. Also note that disputes are included with priority. For security reasons we push as much dispute statements as possible in a block and fill in the rest with backed candidates and bit fields. So far so good.

What bothers me is that I noticed disputes weight is calculated in a different way compared to backed candidates and bitfields. Here is what I mean.

For disputes we call `limit_and_sanitize_disputes` [here](https://github.com/paritytech/polkadot-sdk/blob/f50054cffba4df6b25a16dafdbc66d582e419355/polkadot/runtime/parachains/src/paras_inherent/mod.rs#L445). If the disputes set won't fit in the block it gets reduced until it is small enough. This happens [here](https://github.com/paritytech/polkadot-sdk/blob/f50054cffba4df6b25a16dafdbc66d582e419355/polkadot/runtime/parachains/src/paras_inherent/mod.rs#L445). `limit_and_sanitize_disputes` returns the weight of the reduced set and it is used as 'disputes weights' in `apply_weight_limit` [here](https://github.com/paritytech/polkadot-sdk/blob/f50054cffba4df6b25a16dafdbc66d582e419355/polkadot/runtime/parachains/src/paras_inherent/mod.rs#L445). In nutshell - **disputes are 'sanitized' and then the weight of what's left is calculated**.

However for backed candidates and bitfields the approach is different. Let's see what's done for backed candidates. `apply_weight_limit` gets the whole set (without any sanitization), limits it (if necessary) and returns the weight of the reduced set. And after that `sanitize_backed_candidates` is called ([here](https://github.com/paritytech/polkadot-sdk/blob/f50054cffba4df6b25a16dafdbc66d582e419355/polkadot/runtime/parachains/src/paras_inherent/mod.rs#L590) which potentially reduces this set further ([here](https://github.com/paritytech/polkadot-sdk/blob/f50054cffba4df6b25a16dafdbc66d582e419355/polkadot/runtime/parachains/src/paras_inherent/mod.rs#L924)). So to resume - **backed candidates are weight limited first, weight of the reduced set is calculated and then this set is sanitized and maybe reduced further**.

I wonder if there is a reason for the two approaches I outlined above or this is an oversight? 

Even if it is an oversight it's not a big deal - worst case we will produce underweight blocks. But I think we should either fix this or add a comment explaining it.

**Comment by eskimor** [comment]
For backed candidates, we don't actually expect sanitation to do anything - or at least not much, while for disputes initially before your addition of the `disputes` runtime api we just stuffed in as many disputes as we could and then filtered. There it was absolutely mandatory to calculate the weight after the filtering.

Given that we are now smarter and should not pass disputes and statements to the runtime that will be filtered out, we could probably simplify code a bit more. In general, in my last PR, I did a little bit of cleanup, but there could be done way more.
